src/main/java/vendored/org/apache/dolphinscheduler/common/utils/process/ProcessImplForWin32.java

```diff
@@ -752,13 +752,29 @@ private static void restoreIOEHandleState(WinNT.HANDLE[] stdIOE, Boolean[] inher
                 if (ioRedirectSuccess) {
                     WinBase.PROCESS_INFORMATION pi = new WinBase.PROCESS_INFORMATION();
                     pi.clear();
-
-                    si.dwFlags = WinBase.STARTF_USESTDHANDLES;
+                    si.dwFlags = WinBase.STARTF_USESTDHANDLES | WinBase.STARTF_USESHOWWINDOW;
+                    si.wShowWindow = new WinDef.WORD(WinUser.SW_HIDE);  // hide new console window
                     si.write();
                     final boolean createProcSuccess;
                     final String createProcContext;
                     final int createProcError;
-                    if (isService.get()) {
+
+                    if (username == null) {
+                        createProcContext = "CreateProcess as current user";
+
+                        // From API doc:
+                        // The Unicode version of this function, CreateProcessW, can modify the contents of this string
+                        // Therefore, this parameter cannot be a pointer to read-only memory
+                        // (such as a const variable or a literal string). If this parameter is a constant string,
+                        // the function may cause an access violation.
+                        WTypes.LPWSTR cmdWstr = new WTypes.LPWSTR(cmd);
+                        char[] cmdChars = cmdWstr.getPointer()
+                                .getCharArray(0, cmd.length() + 1);  // +1 terminating null char
+                        WTypes.LPWSTR lpEnvironment = envblock == null ? new WTypes.LPWSTR() : new WTypes.LPWSTR(envblock);
+                        createProcSuccess = Kernel32.INSTANCE.CreateProcessW(null, cmdChars, null, null,
+                                true, new WinDef.DWORD(PROCESS_CREATION_FLAGS), lpEnvironment.getPointer(), path, si, pi);
```

Comment:
where do we get the environment for running as the same user @tilo-chen ?

Reply:
We call ProcessBuilderForWin32.environment() in WindowsExec.createProcess. That initializes the environment, inheriting current process env vars.